packages/hooks/src/useAppHandler/index.ts

```diff
@@ -0,0 +1,437 @@
+
+const runGraphFetch = async (
+  req: {
+    url?: string;
+    userInputs?: Record<string, Record<string, EncreData>>;
+    appPath?: string;
+    context?: Record<string, EncreData>;
+    filter?: ProcessStreamEventFilter;
+  },
+  nodeCallbacks: {
+    [key in string]: {
+      onNodeStart?: (
+        nodeId: string,
+        nodeTitle: string,
+        inputs: Record<string, EncreData>,
+      ) => unknown;
+      onNodeFinish?: (
+        nodeId: string,
+        nodeTitle: string,
+        outputs: Record<string, EncreData>,
+      ) => unknown;
+    };
+  },
+  graphCallbacks: {
+    onDone?: (
+      graphOutput: Record<string, Record<string, EncreData>>,
+    ) => unknown;
+    onError?: (error: string) => unknown;
+  },
+) => {
+  const url: string =
+    req.url ?? process.env.ENCRE_STUDIO_URL ?? 'http://localhost:5127';
+  const apiUrl = `${url}/api/v1/app/run`;
+
+  const res = await fetch(apiUrl, {
+    headers: {
+      'Content-Type': 'application/json',
+    },
+    method: 'POST',
+    body: JSON.stringify({
+      userInputs: req.userInputs ?? {},
+      appPath: req.appPath,
+      context: req.context,
+      filter: req.filter,
+    }),
+  });
+
+  if (res.status !== 200) {
+    const statusText = res.statusText;
+    throw new Error(
+      `The Encre Studio API has encountered an error with a status code of ${res.status} and message ${statusText}`,
+    );
+  }
+
+  const encoder = new TextEncoder();
+  const decoder = new TextDecoder();
+
+  return new ReadableStream({
+    async start(controller) {
+      const onParse = (event: ParsedEvent | ReconnectInterval) => {
+        if (event.type === 'event') {
+          const data = event.data;
+
+          if (data === '[DONE]') {
+            controller.close();
+            return;
+          }
+
+          try {
+            const json = JSON.parse(data);
+
+            // console.log(`runGraphFetch: json: ${JSON.stringify(json)}`);
+            // console.log(
+            //   `nodeCallbacks: ${JSON.stringify(
+            //     Object.keys(nodeCallbacks["ZSJ3g8tL6qsJU6R6c"])
+            //   )}`
+            // );
+
+            if (json['#event'] === 'nodeStart') {
+              const nodeId = json['nodeId'];
+
+              // console.log(`nodeStart: nodeId: ${nodeId}`);
+
+              const onNodeStart = nodeCallbacks[nodeId]?.onNodeStart;
+              const text = onNodeStart?.(
+                json['nodeId'],
+                json['nodeTitle'],
+                json['inputs'],
+              );
+
+              const metadata = {
+                event: 'nodeStart',
+                nodeId: json['nodeId'],
+                nodeTitle: json['nodeTitle'],
+              };
+
+              const encodedText = `metadata: ${JSON.stringify(
+                metadata,
+              )}\ntext: ${JSON.stringify(text)}\n\n`;
+
+              // console.log(`encodedText: ${encodedText}`);
+
+              const queue = encoder.encode(encodedText);
```

Comment:
what is the purpose of encoding an encodedText?

Reply:
same as above. This is for sending the data to the ReadableStream